pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java

```diff
@@ -288,14 +290,35 @@ public static ExtractedSinkDetails validate(SinkConfig sinkConfig, Path archiveP
         }
 
         String sinkClassName;
-        ClassLoader classLoader;
+        final Class<?> typeArg;
+        final ClassLoader classLoader;
         if (!isEmpty(sinkConfig.getClassName())) {
             sinkClassName = sinkConfig.getClassName();
+            // We really don't know if we should use nar class loader or regular classloader
+            ClassLoader jarClassLoader = null;
+            ClassLoader narClassLoader = null;
             try {
-                classLoader = Utils.extractClassLoader(archivePath, functionPkgUrl, uploadedInputStreamAsFile);
+                jarClassLoader = Utils.extractClassLoader(archivePath, functionPkgUrl, uploadedInputStreamAsFile);
             } catch (Exception e) {
-                throw new IllegalArgumentException("Invalid Sink Jar");
             }
+            try {
+                narClassLoader = Utils.extractNarClassLoader(archivePath, functionPkgUrl, uploadedInputStreamAsFile);
+            } catch (Exception e) {
+            }
+            if (jarClassLoader == null && narClassLoader == null) {
+                throw new IllegalArgumentException("Invalid Sink Package");
+            }
+            Class<?> typArg;
+            ClassLoader clsLoader;
+            try {
+                typArg = getSinkType(sinkClassName, narClassLoader);
+                clsLoader = narClassLoader;
+            } catch (Exception e) {
+                typArg = getSinkType(sinkClassName, jarClassLoader);
+                clsLoader = jarClassLoader;
+            }
+            typeArg = typArg;
```

Comment:
this doesn't seem to be right, no?

Reply:
Whats the issue? Also I added some more comments on what I'm doing there.

Reply:
sorry I misread the sentence. I was reading it as `typeArg = typeArg`. I would suggest renaming the arguments. otherwise this part of code is very hard to understand and maintain.

Reply:
renamed